### ISSUE TITLE
Have CMake show a summary of the build options/decisions

### ISSUE THREAD

**Issue by carlosmn** [libgit2/libgit2#3200]
It'd be cool to have CMake show a list of the options and decisions it took and which will affect the build. Stuff like which crypto library we're using, what we're using for talking over the network, what we use for HTTP, zlib and regex.

**Comment by iankronquist** [comment]
I'm not terribly familiar with CMake but I may be able to help. Do you have an example of this in other projects? Do you want this to happen for all build targets or just some specific ones?

**Comment by carlosmn** [comment]
I don't know off-hand any projects which do this in CMake, but e.g. mono or smuxi provide a summary at the end. I was thinking that it would look something like

```
Crypto backend: SecureTransport
cURL support: yes
HTTP stack: http-parser
```

It wouldn't be tied to any target, but to the configuring itself. Towards the end of the config script, there'd be a bunch of code to figure out which one is used and what text to write out.

**Comment by iankronquist** [comment]
Is this the sort of thing you would like to see? I threw this together in 5 minutes and want to make sure I'm on the right track.
https://github.com/libgit2/libgit2/compare/master...iankronquist:cmake-lib-summary/3200

```
Ian@kartal:(build)(master) → $ cmake ..
-- Found Security /System/Library/Frameworks/Security.framework
-- Found CoreFoundation /System/Library/Frameworks/CoreFoundation.framework
-- Could NOT find HTTP_Parser (missing:  HTTP_PARSER_INCLUDE_DIR HTTP_PARSER_LIBRARY) 
-- http-parser was not found or is too old; using bundled 3rd-party sources.
-- checking for module 'libssh2'
--   package 'libssh2' not found
-- LIBSSH2 not found. Set CMAKE_PREFIX_PATH if it is installed outside of the default search path.
-- Found Iconv: -L/usr/lib -liconv
-- Performing Test IS_FVISIBILITY=HIDDEN_SUPPORTED
-- Performing Test IS_FVISIBILITY=HIDDEN_SUPPORTED - Success
-- Performing Test IS_WSTRICT-ALIASING=2_SUPPORTED
-- Performing Test IS_WSTRICT-ALIASING=2_SUPPORTED - Success
-- SSL: 
-- SSH: 
-- GSSAPI: 
-- ICONV: -L/usr/lib -liconv
-- HTTP parser: 
-- Configuring done
...
```

**Comment by carlosmn** [comment]
It's a good start, but I'm not so worried about specifying the library paths (we can look that up if we really need them) but an overview of the capabilities which libgit2 was built with. There's also a few places where we have to be a bit smart about it.

- We may use our copy of zlib or the sytem's, and we should say which one, rather than flags. The same goes for http-parser. If the use the built-in one, the flags are empty, which doesn't say a lot.
- Crypto libraries may exist at two levels.
  - There's what we use for SHA-1 (which may be built-in, OpenSSL, CommonCrypto or WinCNG depending on the OS and available libraires)
  - and there's what we use to speak HTTPS (which can be nothing, OpenSSL, SecureTransport or WinHTTP) again depending on the OS and what's available.

We also really should stay away from mentioning SSL specifically, as we never use that (if anything, we use TLS, but that's also probably more specific than it needs to be). I'd split that up into two lines named "SHA-1" and "HTTPS" or similar.

For the SHA-1 implementation, the selection happens starting at line 232. It might be easiest to set a string variable in each of those blocks which provides the description. The TLS/HTTPS selection is a bit all over the place, but we do have the variables `WINHTTP`, `OPENSSL_FOUND` and `SECURITY_FOUND`, which correspond to WinHTTP, OpenSSL and SecureTransport. Note that we use SecureTransport if it's detected, regardless of whether we've also found OpenSSL.

For the others (iconv, gssapi, curl, libssh2) we can stick to yes/no, I think.

**Comment by iankronquist** [comment]
Thank you. Those details are much appreciated.

**Comment by 0x41111111** [comment]
Hi, I've submitted a PR for this issue. Most of the changes are down toward the end of the file, with the SHA-1 selection stuff being modified to set a variable depending on the implementation in use.

Example:
```
-- Build options/dependencies:
-- HTTP parser: bundled
-- zlib: system-provided
-- libiconv: no
-- libcurl: no
-- gssapi: no
-- libssh2: no
-- Crypto library (SHA-1): OpenSSL
-- Crypto library (HTTPS): OpenSSL
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mobile/projects/github/libgit2
```

**Comment by carlosmn** [comment]
Fixed via #4283
